### PR TITLE
Automate apache-bench benchmarking

### PR DESCRIPTION
**Pull request by nskool**
## Description
This PR adds functionality within tests to automate torchserve benchmarking that runs apache-bench latency tests on different ec2 instances. It aims to introduce the following features:
1. Use the current ‘serve’ folder closure to run benchmarks on several ec2 instances i.e. changes can be made to any part of current working torchserve folder, and benchmarks can be triggered at once. 
2. Generate the docker image once, re-use it across multiple instances for benchmarking.
3. Switchable ‘benchmark-engine’ e.g., apache-bench/locust.
4. Correctness checks i.e. supplied benchmark config format/versions of different packages.
5. Each model benchmark across any CPU/GPU instance executes in parallel to other models.
6. Run each model benchmark individually or all at once.
7. Allow for better specification of benchmarking config using fewer config files.

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [x] `python test/benchmark/run_benchmark.py`

- UT/IT execution results
Sample output:
```
[gw0] PASSED tests/test_vgg11.py::test_vgg11_benchmark[c5.4xlarge]

========================================================================================================================== PASSES ===========================================================================================================================
_____________________________________________________________________________________________________________ test_vgg11_benchmark[c5.4xlarge] ______________________________________________________________________________________________________________
[gw0] darwin -- Python 3.6.12 /Users/nikhilsk/nskool/bvenv/bin/python
-------------------------------------------------------------------------------------------------------------------- Captured log setup ---------------------------------------------------------------------------------------------------------------------
INFO     tests.utils:conftest.py:32 Setting up docker image to be used
INFO     tests.utils:__init__.py:245 docker content:
INFO     tests.utils:__init__.py:246 {'gpu': {'docker_tag': 'gpu-dev-image', 'cuda_version': 'cu102'}, 'cpu': {'docker_tag': 'cpu-dev-image'}}
INFO     tests.utils:__init__.py:276 Docker YAML config validation successful.
INFO     tests.utils:__init__.py:58 Dev image build successful:  torchserve-benchmark:gpu-dev-image
INFO     tests.utils:__init__.py:90 Dev image push to ECR successful.
INFO     tests.utils:__init__.py:58 Dev image build successful:  torchserve-benchmark:cpu-dev-image
INFO     tests.utils:__init__.py:90 Dev image push to ECR successful.
INFO     tests.utils:conftest.py:167 Created instance: TS Benchmark ec2-key-name-448
INFO     tests.utils:conftest.py:192 Instance ip_address: 44.234.253.130
INFO     tests.utils:conftest.py:194 Connecting to ubuntu@44.234.253.130
--------------------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------------------
INFO     tests.utils:test_vgg11.py:28 Validating yaml contents
INFO     tests.utils:__init__.py:236 Benchmark YAML config validation successful.
INFO     tests.utils:test_vgg11.py:30 True
INFO     tests.utils:__init__.py:126 Dev image pull from ECR successful.
================================================================================================================== short test summary info ==================================================================================================================
PASSED tests/test_vgg11.py::test_vgg11_benchmark[c5.4xlarge]
========================================================================================================= 1 passed, 8 warnings in 479.49s (0:07:59) =========================================================================================================
```

- Logs
Sample benchmark output:

### Benchmark report

**vgg11 | eager_mode | c5.18xlarge | batch size 1**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 2.05 | 47419 | 54745 | 58781 | 48852.156 | 0.0 | 589.16 | 709.42 | 709.42 | 1905.05 | 1904.91 | 44589.48 | 1.09 | 

**vgg11 | eager_mode | c5.18xlarge | batch size 8**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 8.11 | 12205 | 13162 | 14772 | 12334.135 | 0.0 | 3431.05 | 3525.94 | 3525.94 | 3872.42 | 3872.04 | 7958.16 | 53.27 | 

**vgg11 | eager_mode | c5.18xlarge | batch size 4**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 5.55 | 17891 | 18936 | 19965 | 18017.484 | 0.0 | 2304.79 | 2412.98 | 2412.98 | 2820.51 | 2820.24 | 14423.28 | 52.17 | 

**vgg11 | eager_mode | c5.18xlarge | batch size 2**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 3.51 | 28732 | 29900 | 30531 | 28520.545 | 0.0 | 748.97 | 1431.84 | 1431.84 | 2226.96 | 2226.79 | 25045.02 | 49.16 | 

**vgg11 | scripted_mode | c5.18xlarge | batch size 1**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 2.06 | 48058 | 50794 | 51760 | 48618.091 | 0.0 | 874.51 | 1012.23 | 1012.23 | 1900.22 | 1900.11 | 44363.84 | 1.07 | 

**vgg11 | scripted_mode | c5.18xlarge | batch size 4**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 5.50 | 18055 | 19159 | 19844 | 18171.083 | 0.0 | 2230.75 | 2316.4 | 2316.4 | 2846.7 | 2846.34 | 14550.68 | 51.29 | 

**vgg11 | scripted_mode | c5.18xlarge | batch size 3**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 1 | 4.51 | 22138 | 23074 | 23792 | 22165.721 | 0.1 | 1804.03 | 2160.02 | 2160.02 | 2597.17 | 2597.01 | 18563.88 | 50.2 | 

**vgg11 | scripted_mode | c5.18xlarge | batch size 2**
 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
 | AB | vgg11 | 100 | 1000 | 0 | 3.47 | 28765 | 29849 | 30488 | 28781.227 | 0.0 | 1576.24 | 1758.28 | 1758.28 | 2249.52 | 2249.34 | 25210.43 | 46.77 | 

## Checklist:

- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] New and existing unit tests pass locally with these changes?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by msaroufim** [comment]
I tried running this and got the below error 

`
(servetest) ubuntu@ip-172-31-1-60:~/serve$ python test/benchmark/run_benchmark.py
Running tests from directory: /home/ubuntu/serve/test/benchmark/tests
Running pytest
ERROR: usage: run_benchmark.py [options] [file_or_dir] [file_or_dir] [...]
run_benchmark.py: error: unrecognized arguments: -n=4
  inifile: None
  rootdir: /home/ubuntu/serve
`

Also would be worthwhile to add a reference for how to run this in this page https://github.com/pytorch/serve/tree/master/benchmarks

Basically 
1. pip install -r test/benchmark/requirements.txt
2. python test/benchmark/run_benchmark.py

With some note that this is an expensive operation running jobs on real servers

**Comment by nskool** [comment]
> I tried running this and got the below error
> 
> `(servetest) ubuntu@ip-172-31-1-60:~/serve$ python test/benchmark/run_benchmark.py Running tests from directory: /home/ubuntu/serve/test/benchmark/tests Running pytest ERROR: usage: run_benchmark.py [options] [file_or_dir] [file_or_dir] [...] run_benchmark.py: error: unrecognized arguments: -n=4 inifile: None rootdir: /home/ubuntu/serve`
> 
> Also would be worthwhile to add a reference for how to run this in this page https://github.com/pytorch/serve/tree/master/benchmarks
> 
> Basically
> 
>     1. pip install -r test/benchmark/requirements.txt
> 
>     2. python test/benchmark/run_benchmark.py
> 
> 
> With some note that this is an expensive operation running jobs on real servers

Thanks Mark, yes, definitely! That's one of the pending items, I left the ' Have you made corresponding changes to the documentation?' unticked, and will update the PR soon with some more changes to address this.

File: docker/Dockerfile.dev [hunk 1]
```diff
@@ -52,7 +52,7 @@ ARG CUDA_VERSION
 RUN if [ "$MACHINE_TYPE" = "gpu" ]; then export USE_CUDA=1; fi \
     && git clone https://github.com/pytorch/serve.git \
     && cd serve \
-    && git checkout ${BRANCH_NAME} \
+    && git checkout --track origin/release_0.4.0 \
```
Comment (dhanainme) [inline docker/Dockerfile.dev:55]: Fix branch name `release_0.4.0`

File: test/benchmark/run_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.DEBUG)
+LOGGER.addHandler(logging.StreamHandler(sys.stdout))
+
+
+def main():
+    # Run this script from the root directory 'serve', it changes directory below as required
+    os.chdir(os.path.join(os.getcwd(), "test", "benchmark"))
+
+    execution_id = f"ts-benchmark-run-{str(uuid.uuid4())}"
```
Comment (dhanainme) [inline test/benchmark/run_benchmark.py:23]: Suggestion. : Date prefix if it would help.

File: test/benchmark/tests/utils/__init__.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+DEFAULT_DOCKER_DEV_ECR_TAG = "dev-image"
+DEFAULT_DOCKER_DEV_ECR_REPO_TAG = f"{DEFAULT_DOCKER_DEV_ECR_REPO}:{DEFAULT_DOCKER_DEV_ECR_TAG}"
+ECR_REPOSITORY_URL = "{}.dkr.ecr.{}.amazonaws.com/{}"
+
+GPU_INSTANCES = ["p2", "p3", "p4", "g2", "g3", "g4"]
+
+# DLAMI with nVidia Driver ver. 450.119.03 (support upto CUDA 11.2), Ubuntu 18.04
+AMI_ID = "ami-0ff137c06803a8bb7"
+# AMI_ID = "ami-0198925303105158c", with apache2-utils installed
+
```
Comment (dhanainme) [inline test/benchmark/tests/utils/__init__.py:32]: Can this be made a part of config ?

File: test/benchmark/tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+    request.addfinalizer(delete_ssh_keypair)
+
+    params = {
+        "KeyName": ec2_key_name,
+        "ImageId": ec2_instance_ami,
+        "InstanceType": ec2_instance_type,
+        "IamInstanceProfile": {"Name": ec2_instance_role_name},
+        "TagSpecifications": [
+            {"ResourceType": "instance", "Tags": [{"Key": "Name", "Value": f"TS Benchmark {ec2_key_name}"}]},
```
Comment (dhanainme) [inline test/benchmark/tests/conftest.py:178]: Can this be a config ?

**Review by dhanainme — COMMENTED** [review]
Also consider emitting metrics to alarm on for failure to launch instance / Non 0 application exits. This will ensure we have some thing to alarm on for EC2 Instances not terminated

File: test/benchmark/tests/utils/report.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+class MarkdownTable:
+    def __init__(self):
+        self.table_content = ""
+
+    def create_two_column_table(self, header_1: str, header_2: str):
+        self.table_content += f"| {header_1} | {header_2} |\n | --- | --- |"
+
+    def add_row_to_two_column_table(self, column_1_content: str, column_2_content: str):
+        column_1_rows = column_1_content.split("\n")
+        column_2_rows = column_2_content.split("\n")
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/tests/utils/report.py:32]: I'd considered this, and will probably use it in the next iteration to make the report more complex.

File: test/benchmark/tests/conftest.py (hunk 4 above)
Comment (nskool) [reply test/benchmark/tests/conftest.py:178]: Do you mean it should be possible to chose a naming pattern for the instances? I think that makes sense, added to future improvements in this issue: https://github.com/pytorch/serve/issues/1067

File: test/benchmark/tests/utils/__init__.py (hunk 3 above)
Comment (nskool) [reply test/benchmark/tests/utils/__init__.py:32]: The objective is that __init__.py functions as a config. However, yes, it makes sense given the other configs used. Noted here: https://github.com/pytorch/serve/issues/1067

File: test/benchmark/run_benchmark.py (hunk 2 above)
Comment (nskool) [reply test/benchmark/run_benchmark.py:23]: The uuid4() is based on the current timestamp as well: https://docs.python.org/3/library/uuid.html#uuid.UUID.fields. This can be made more human readable, making a note. https://github.com/pytorch/serve/issues/1067

File: docker/Dockerfile.dev (hunk 1 above)
Comment (nskool) [reply docker/Dockerfile.dev:55]: Thanks, fixed.

File: test/benchmark/run_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+from invoke import run
+from invoke.context import Context
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.DEBUG)
+LOGGER.addHandler(logging.StreamHandler(sys.stdout))
+
+
+def main():
+    # Run this script from the root directory 'serve', it changes directory below as required
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/run_benchmark.py:20]: Yes, the README does capture this: https://github.com/pytorch/serve/blob/668a4d178cb6975fdbde05aa19ca335525f6527e/test/benchmark/README.md

File: test/benchmark/tests/utils/s3.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    :param folder_path: path of the folder that must be uploaded
+    :param s3_uri: desired s3 path for the folder to reside in
+    :param connection: if provided, s3 command will be run on the ec2 instance
+    :return: True if uploaded succeeded else False
+    """
+
+    if connection:
+        run_out = connection.run(f"aws s3 cp --recursive {folder_path}/ {s3_uri}/")
+    else:
+        run_out = run(f"aws s3 cp --recursive {folder_path}/ {s3_uri}/")
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/tests/utils/s3.py:31]: Using boto3 for both remote instance and local instance would be a little tedious. Since the client is already authenticated, this is simpler, and keeps the experience same when running locally, or on the ec2 instance. But I've noted this, and can be introduced in a later iteration for robustness.

**Review by dhanainme — APPROVED** [review]

**Comment by msaroufim** [comment]
Thank you @nskool this is a fascinating PR

Just confirming that the timeout error is happening both on an AWS EC2 instance and my local mac

https://gist.github.com/msaroufim/f88b8cdfdf4d62ba451014b02fcf430e

In `__init__.py` 

```
DEFAULT_REGION = "us-west-2"
IAM_INSTANCE_PROFILE = "EC2Admin"
S3_BUCKET_BENCHMARK_ARTIFACTS = "s3://benchmarktest"
```
and output of 


```(test) marksaroufim@marksaroufim-mbp utils % aws sts get-caller-identity         
{
    "UserId": "AIDAUVI2STJOZSKLHMREA",
    "Account": "320567679581",
    "Arn": "arn:aws:iam::320567679581:user/marksaroufim"
}
```

File: test/benchmark/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+* Adjust the following global variables to your preference in the file serve/test/benchmark/tests/utils/__init__.py <br>
+-- IAM_INSTANCE_PROFILE :this role is attached to all ec2 instances created as part of the benchmarking process. Create this as described [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#create-iam-role). Default role name is 'EC2Admin'.<br>
+-- S3_BUCKET_BENCHMARK_ARTIFACTS :all temporary benchmarking artifacts including server logs will be stored in this bucket: <br>
+-- DEFAULT_DOCKER_DEV_ECR_REPO :docker image used for benchmarking will be pushed to this repo <br>
+
+*The following steps assume that the current working directory is serve/.*
+
+1. Create or use any python virtual environment
+```
+python3 -m venv bvenv
```
Comment (msaroufim) [inline test/benchmark/README.md:24]: apt-get install python3-venv
pip install cryptography==2.5

File: test/benchmark/README.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+source bvenv/bin/activate
+```
+2. Install requirements for the benchmarking 
+```
+pip install -r test/benchmark/requirements.txt
+```
+3. Make sure that you've setup AWS account correctly
+```
+aws sts get-caller-identity
```
Comment (msaroufim) [inline test/benchmark/README.md:33]: add iam:roles

Comment (msaroufim) [inline test/benchmark/README.md:33]: When setting up base instance make sure to set IAM role to EC2ADmin with the permissions above

Maybe add
* SSH port and with all inbound access

**Review by msaroufim — APPROVED** [review]
LGTM - this is an amazing feature to have and will save us so much time

Here's the S3-tree that stores the report https://gist.github.com/msaroufim/9deeb1d4ddfbd1e57d2d6f05ae9637d9#file-tree-txt-L7377

And the console output confirming VGG-16 passed

Some minor comments to address in README otherwise looks great!

```
============================================ short test summary info ==============================================
PASSED tests/test_vgg16.py::test_vgg16_benchmark[p3.8xlarge]
SKIPPED [1] tests/test_vgg11.py:26: unconditional skip
SKIPPED [1] tests/test_mnist.py:26: unconditional skip
SKIPPED [1] tests/test_bert.py:28: unconditional skip
SKIPPED [1] tests/test_fastrcnn.py:26: unconditional skip
============================== 1 passed, 4 skipped, 12 warnings in 1649.14s (0:27:29) ==============================
```

**Comment by msaroufim** [comment]
Logs on s3 look good: https://gist.github.com/msaroufim/9deeb1d4ddfbd1e57d2d6f05ae9637d9#file-tree-txt-L7377
Logs on console look good
Great feature and looks good to me - just some minor updates to make on docs and we should merge this

File: test/benchmark/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,10 @@
+pytest
+pytest-xdist
+boto3
+pytest-timeout
+pytest-rerunfailures
+retrying
+fabric
+black
+gitpython
+docker
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/requirements.txt]: Updated

File: test/benchmark/tests/test_vgg11.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
+import os
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/tests/test_vgg11.py:1]: That is correct for now, this can be structured better in phase 2. (Pt. 9 here: https://github.com/pytorch/serve/issues/1067)

File: test/benchmark/README.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+Check out a sample vgg11 model config at the path: `tests/suite/vgg11.yaml`
+
+### Setup
+
+* Ensure you have access to an AWS account i.e. [setup](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html) your environment such that awscli can access your account via either an IAM user or an IAM role. An IAM role is recommended for use with AWS. For the purposes of testing in your personal account, the following managed permissions should suffice: <br>
+-- [AmazonEC2ContainerRegistryFullAccess](https://console.aws.amazon.com/iam/home#policies/arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess) <br>
+-- [AmazonEC2FullAccess](https://console.aws.amazon.com/iam/home#policies/arn:aws:iam::aws:policy/AmazonEC2FullAccess) <br>
+-- [AmazonS3FullAccess](https://console.aws.amazon.com/iam/home#policies/arn:aws:iam::aws:policy/AmazonS3FullAccess) <br>
+* [Create](https://docs.aws.amazon.com/cli/latest/reference/ecr/create-repository.html) an ECR repository with the name “torchserve-benchmark” in the us-west-2 region
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/README.md:13]: Updated, thanks

File: test/benchmark/README.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+```
+INSTANCE_TYPES_TO_TEST = ["p3.8xlarge"]
+```
+5. The automation scripts uses the ts-config from the following location: `benchmarks/config.properties`. Make changes to this file in the current local folder to use this across all the runs.
+6. Finally, start the benchmark run as follows (run this a pseudo shell such as tmux or screen, as this is a long-running script):
+```
+python test/benchmark/run_benchmark.py
+```
+7. To start test for a particual model, modify the `pytest_args` list in run_benchmark.py to include `["-k", "vgg11"]`, if that particular model is vgg11
+8. For generating benchmarking report, modify the argument to function `generate_comprehensive_report()` to point to the s3 bucket uri for the benchmark run. Run the script as:
```
[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/README.md:45]: Will be adding this in phase 2: (Pt 6,7 here: https://github.com/pytorch/serve/issues/1067)

[the remark this answers is not in the document]
Comment (nskool) [reply test/benchmark/README.md:45]: That is correct - step 8 needs to run after 6, and will be called automatically after step 6 (tracked here: https://github.com/pytorch/serve/issues/1067)

File: test/benchmark/README.md (hunk 8 above)
Comment (nskool) [reply test/benchmark/README.md:24]: Updated, leaving out the 'cryptography==2.5' line since that's an environment error not specific to the requirements on this PR.

File: test/benchmark/README.md (hunk 9 above)
Comment (nskool) [reply test/benchmark/README.md:33]: Added, thanks.

Comment (nskool) [reply test/benchmark/README.md:33]: Thanks, it wouldn't be necessary for every user to use EC2Admin as an ec2-attached-iam-role, user can chose their own method for the first setup. So I'd leave that part out.

Updated the readme regarding ssh port.
